engine/src/main/java/org/pentaho/di/trans/steps/webservices/wsdl/Wsdl.java

```diff
@@ -0,0 +1,401 @@
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.trans.steps.webservices.wsdl;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+
+import javax.wsdl.Binding;
+import javax.wsdl.Definition;
+import javax.wsdl.Operation;
+import javax.wsdl.Port;
+import javax.wsdl.PortType;
+import javax.wsdl.Service;
+import javax.wsdl.WSDLException;
+import javax.wsdl.extensions.ExtensionRegistry;
+import javax.wsdl.factory.WSDLFactory;
+import javax.wsdl.xml.WSDLLocator;
+import javax.wsdl.xml.WSDLReader;
+import javax.xml.namespace.QName;
+
+import org.apache.http.auth.AuthenticationException;
+import org.pentaho.di.core.HTTPProtocol;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.exception.KettleStepException;
+import org.pentaho.di.core.logging.LogChannel;
+import org.pentaho.di.core.xml.XMLHandler;
+import org.w3c.dom.Document;
+
+/**
+ * Wsdl abstraction.
+ */
+public final class Wsdl implements java.io.Serializable {
+  private static final long serialVersionUID = 1L;
+  private Port _port;
+  private final Definition _wsdlDefinition;
+  private final Service _service;
+  private final WsdlTypes _wsdlTypes;
+  private HashMap<String, WsdlOperation> _operationCache;
+  private URI wsdlURI = null;
+
+  /**
+   * Loads and parses the specified WSDL file.
+   *
+   * @param wsdlURI
+   *          URI of a WSDL file.
+   * @param serviceQName
+   *          Name of the service in the WSDL, if null default to first service in WSDL.
+   * @param portName
+   *          The service port name, if null default to first port in service.
+   */
+  public Wsdl( URI wsdlURI, QName serviceQName, String portName ) throws AuthenticationException {
+    this( wsdlURI, serviceQName, portName, null, null );
+  }
+
+  public Wsdl( URI wsdlURI, QName serviceQName, String portName, String username, String password ) throws AuthenticationException {
+
+    this.wsdlURI = wsdlURI;
+    try {
+      _wsdlDefinition = parse( wsdlURI, username, password );
+    } catch ( AuthenticationException ae ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)